stream_alert/shared/rule.py

```diff
@@ -0,0 +1,187 @@
+See the License for the specific language governing permissions and
+limitations under the License.
+"""
+from stream_alert.shared import LOGGER
+from stream_alert.shared.stats import time_rule
+
+
+class RuleInvalid(Exception):
+    """Exeception to raise for any errors with invalid rules"""
+
+
+def rule(**opts):
+    """Decorator to be used to register a rule"""
+    def decorator(rule_func):
+        """Rule decorator logic that returns instance of Rule"""
+        return Rule(rule_func, **opts)
+    return decorator
+
+
+def disable(rule_instance):
+    """Decorator to be used for disabling a rule"""
+    Rule.disable(rule_instance.rule_name)
+    return rule_instance
+
+
+class Rule(object):
+    """Rule class to handle processing"""
+    _rules = {}
+
+    def __init__(self, func, **kwargs):
+        self.func = func
+        self.rule_name = func.__name__
+        self.logs = kwargs.get('logs')
+        self.outputs = kwargs.get('outputs')
+        self.matchers = kwargs.get('matchers')
+        self.datatypes = kwargs.get('datatypes')
+        self.req_subkeys = kwargs.get('req_subkeys')
+        self.initial_context = kwargs.get('context')
+        self.context = None
+        self.disabled = False
+
+        if not (self.logs or self.datatypes):
+            raise RuleInvalid(
+                "Invalid rule [{}] - rule must have either 'logs' "
+                "or 'datatypes' declared'".format(self.rule_name)
+            )
+
+        if self.rule_name in Rule._rules:
+            raise RuleInvalid('Rule [{}] already defined'.format(self.rule_name))
+
+        Rule._rules[self.rule_name] = self
+
+    def __str__(self):
+        return '<Rule: {}; outputs: {}; disabled: {}>'.format(
+            self.rule_name,
+            self.outputs,
+            self.disabled
+        )
+
+    def __repr__(self):
+        return self.__str__()
+
+    def check_matchers(self, record):
+        """Run any rule matchers against the record
+
+        Args:
+            record (dict): Record that this rule should be run against
+
+        Returns:
+            bools: True if all matchers apply to this record, False otherwise
+        """
+        if not self.matchers:
+            return True
+
+        return all(Matcher.process(matcher_name, record) for matcher_name in self.matchers)
+
+    @time_rule
+    def process(self, record):
+        """Process will call this rule's function on the passed record
+
+        Args:
+            record (dict): Record that this rule should be run against
+
+        Returns:
+            bool: True if this rule triggers for the passed record, False otherwise
+        """
+        try:
+            # The initial_context object must be copied. This avoids
+            # bleed over from other runs of the rule using the same
+            # context object
+            if self.initial_context is not None:
+                self.context = self.initial_context.copy()
```

Comment:
`dict.copy()` makes a shallow copy:

```python
context = {'a': {'b': 'c'}}
result = context.copy()
result['a']['b'] = 3
print context
# {'a': {'b': 3}}
```

As far as I know, `context` can be a nested dict, so we should use `copy.deepcopy` instead

Reply:
good suggestion - I'll update this :) as a note, this is addressing a previously existing bug that didn't copy the context at all. better to get it right now than on the 3rd try 🤣

Reply:
Yeah, it's always funny how a refactor resurfaces a bunch of code which didn't necessarily change but gets reviewed again. It's probably for the best lol thanks for doing that, even though it wasn't introduced by you